### ISSUE TITLE
[Bug] NiftiLabelsMasker does not ignore background

### ISSUE THREAD

**Issue by 0nu** [nilearn/nilearn#2060]
I am using `NiftiLabelsMasker` to label subject data into atlasses. Despite of being transformed to MNI space, the atlases usually comprise a larger area / volume than the subject data. 
Obviously, I 'd like the areas of no subject data to be excluded from the label mean-computation, so that each label's value is the mean of the data, and not the mean of a mixture of data and background. According to my tests, this is not possible with `NiftiLabelsMasker`. 
Setting `background_label=` does not change this behavior, neither does `mask_img`.  

I am happy to try to clarify this issue or provide more information if something is missing. 

**What version of Nilearn are you using?** 
nilearn.input_data.NiftiLabelsMasker 0.5 .2 with python 3.5. Same behavior with git version of nilearn.

**What were you trying to do?** 
NiftiLabelsMasker.transform() with an atlas that comprises larger area as the subject data  

**What did you expect will happen?** 
I 'd expected the background areas to be excluded from the mean computation of the corresponding label.  

**What actually happened?** 
Background areas were included in computation.  

**Proposed solution**  
1. Pass `mask_img` to `regions.signal_extraction.img_to_signals_labels()`, as already proposed in  #1257. This would be possible by changing `_ExtractionFunctor`. This does not solve the problem if no `mask_img` is provided. 
2. Change `img_to_signals_labels()` to remove `background_label` from mean computation.  

**Code example for problem** 
```
"""Check effect of missing data on NiftiLabelsMasker label values"""
from nilearn.input_data import NiftiLabelsMasker
import nibabel as nib
import numpy as np

background_label = 0.
dims = (10, 10, 10)
# random affine for the Niimg object
affine = np.array([[-1, 0, 0, 90],
                   [0, 1, 0, -126],
                   [0, 0, 1, -72],
                   [0, 0, 0, 1]])

# create atlas img with one label
labels_data = np.zeros(dims)
labels_data[0:5, 0:5, 0:5] = 1
labels_img = nib.Nifti1Image(labels_data, affine, nib.Nifti1Header())

# create data
data = np.zeros(dims + (3,))
data[:] = background_label
data[0:5, 0:5, 0:5, 1] = 1  # second slice: 1s at same area as labels_img
data[0:3, 0:3, 0:3, 2] = 1  # third slice:  1s in subset of the label area

# build image from data array
img = nib.Nifti1Image(data, affine, nib.Nifti1Header())

masker = NiftiLabelsMasker(labels_img=labels_img, background_label=background_label)

print("Expected labels with background removed from data: {}".format([[background_label], [1.], [1.]]))
print("NiftiLabelsMasker results without mask: {}".format(masker.fit_transform(img).tolist()))  # [[0.], [1.], [.216]]

# the same with mask_img
mask = np.zeros(dims + (3,))
mask[data != background_label] = 1  # set the mask to data
mask = np.max(mask, axis=3)
mask_img = nib.Nifti1Image(mask, affine, nib.Nifti1Header())

masker_with_mask = NiftiLabelsMasker(labels_img=labels_img,
                                     background_label=background_label,
                                     mask_img=mask_img)
print("NiftiLabelsMasker results with mask: {}".format(
    masker_with_mask.fit_transform(img).tolist()))  # [[0.], [1.], [.216]]

```

**Comment by jeromedockes** [comment]
the results you get are the expected behaviour:
```
>>> data[labels_data == 1].mean(axis=0)
array([0.   , 1.   , 0.216])
```
(where `data` and `labels_data` are defined as in your script). the output of the labels masker has shape (time points x number of labels) and the entry i,j is the mean of voxels from region j at time point i. in your case there is only one region and 3 time points (the last dimension of your 4d nifti image), so the shape is (3 x 1). if you need more explanations, please prefer the neurostars forum (https://neurostars.org/) for usage questions and to ask for help

**Comment by 0nu** [comment]
Thank you for taking the time to respond Jerome!
The point I was trying to make was, that the label from the last volume should also have a value of `1` and not `.0216`, as the background label (`0`) should not be taken into the mean.

**Comment by jeromedockes** [comment]
values in voxels that are marked as background in the labels_img of the masker are not taken into account. zeros in the transformed image `img` in your script are treated as the other values. values in img are not region labels; the region labels are in the labels_img provided when creating the NiftiLabelsMasker. as you can see from the comment above https://github.com/nilearn/nilearn/issues/2060#issuecomment-493445453 the masker gives the correct result, ie the mean of image values in the region for which labels_img == 1

**Comment by 0nu** [comment]
A given mask to `NiftiLabelsMasker` is not passed to `signal_extraction.img_to_signals_labels()`. 

Sorry about my confusion about the `background_label`, but the mask issue still exists IMO. This makes this issue a duplicate of #1257 then...
``` 
mask_3rdslice = np.zeros(dims)
mask_3rdslice[data[...,2] != background_label] = 1  # set the mask to data of 3rd slice
mask_3rdslice_img = nib.Nifti1Image(mask, affine, nib.Nifti1Header())
masker_with_3rdslice_mask = NiftiLabelsMasker(labels_img=labels_img,
                                              background_label=background_label,
                                              mask_img=mask_3rdslice_img)
print("NiftiLabelsMasker results with 3rd slice mask: {}".format(
    masker_with_3rdslice_mask.fit_transform(img).tolist()))  # [[0.], [1.], [.216]]
print(data[np.logical_and(labels_data == 1, mask_3rdslice == 1)].mean(axis=0)) # [0. 1. 1.]
```

**Comment by jeromedockes** [comment]
> print(data[np.logical_and(labels_data == 1, mask_3rdslice == 1)].mean(axis=0)) # [0. 1. 1.]

you're not using `img` here. this:
```
>>> data[labels_data == 1].mean(axis=0)
array([0.   , 1.   , 0.216])
```
ie the mean of the image data inside the region mask, is what the masker is supposed to compute, and it is indeed the result you find with it.

**Comment by jeromedockes** [comment]
@0nu do you have any more questions or can we close this?

**Comment by jeromedockes** [comment]
nilearn displays the expected behaviour in the examples above so closing this issue
